### PR TITLE
Primitive Cleaning

### LINKED ISSUE #564: Refactor primitives into a submodule
The primitives file is pretty big and I'm wondering whether, organizationally, if it might make sense to refactor it into its own module and split up the various traits and primitive implementations.

Marked as Needs Triage pending further discussion

### PR DESCRIPTION
**Pull request by rachitnigam**
Taking a shot at reorganizing the primitives.rs file.

Fixes #564.

### CODE REVIEW

**Comment by sampsyo** [comment]
This looks awesome! I hope it makes things less repetitive.

**Comment by rachitnigam** [comment]
This is blocked on #583, #575, and #586 (the last two touch the primitive tests). We can continue working on it once they are merged.

**Comment by rachitnigam** [comment]
Rebased on #596 and have it `StdReg` working. @EclecticGriffin can you help me figure out how to do serialization in a way that we can still use trait objects (the normal trick of having a function `fn serialize<S>` won't work because trait objects don't allow that.) 

There should be some way of returning a simpler data structure that can be serialized in the environment.

**Comment by EclecticGriffin** [comment]
Yeah, I'll do some investigating

**Comment by EclecticGriffin** [comment]
@rachitnigam see how you feel about what I just pushed. Does add a new dependency, which isn't ideal, so we could decide to replace the relevant code with my iterator crimes from the previous approach, though I'm hopeful this will scale a bit better if we decide to introduce higher dimensional memories.

**Comment by rachitnigam** [comment]
Oh, this is a super neat trick. I didn't know you could do `impl dyn ...` for trait objects. I'll say that it's better to have your serialization crimes in rather than a new dependency.

**Comment by rachitnigam** [comment]
Added more primitives. The remaining primitives need to implemented by hand for now. The `std_not`, `std_slice`, and `std_slice` primitives cannot be implemented for a stupid reason: we can't have a port named `in` because `in` is a Rust keyword. Raw identifiers don't work with the macro because `stringify!(...)` generates the `r#` prefix. Changing the port name in the standard library is a chore because Dahlia also refers to the names of the input ports.

**Comment by EclecticGriffin** [comment]
Yeah I'll swap in my serialization crimes in a few

**Comment by EclecticGriffin** [comment]
@rachitnigam I might've solved the macro issue. Take a look

**Comment by rachitnigam** [comment]
Nice hack!!!

**Comment by rachitnigam** [comment]
This is ready to merge once @almathaler @ai-cu and @yn224 get a chance to take a look.

File: interp/src/primitives/combinational.rs
Comment (almathaler) [inline interp/src/primitives/combinational.rs:20]: what's this `ir::Binding` thing? I'm not sure I've seen it before, is it new?

File: interp/src/stk_env.rs
Comment (almathaler) [inline interp/src/stk_env.rs:875]: why `from` and not `try_from_init`? I've just always been using `try_from_init`, is that not always the best choice?

File: interp/src/primitives/combinational.rs
Comment (EclecticGriffin) [reply interp/src/primitives/combinational.rs:20]: You can find the definition in the calyx portion of the repo, but it's just a smallvec tuple containing parameter Ids associated with their u64 values.

File: interp/src/stk_env.rs
Comment (EclecticGriffin) [reply interp/src/stk_env.rs:875]: `try_from_init` was renamed to `from` It's the same method

File: interp/src/values.rs
Comment (almathaler) [inline interp/src/values.rs:82]: Never mind the above comment, I see this is just a re-name

**Review by almathaler — APPROVED** [review]
this all looks good!

**Pull request merged**
